### DISCUSSION TITLE
Regarding Turn by turn navigation implementation.

### DISCUSSION THREAD

**Discussion by jesussmile** [liodali/osm_flutter#136] [Show and tell]
Here's what I have done. Downloaded all the JSON data that contains the API then using some algorithm calculated the latitude and longitude values for all the turns and directions and compared them with my current position,  I kept a certain value like 50 or 100 meters to warn the users about changes in direction. I streamed my position to always be in the center of the map with heading and speed. With heading, I can change the orientation of the map to show the road straight ahead. I get a very rough turn-by-turn implementation. Do you think this approach is okay? In the future, I would like to see how you have implemented the polylines and try to compute the directions offline. 
It's funny how the arrow is dragging my car!  😄 

https://user-images.githubusercontent.com/11044978/127325020-c21c332d-0b3f-4fa4-a005-5b996c773df9.mp4

**Comment by liodali** [comment]
yeah, it's  good start to add turn by turn navigation 
for calculation did you calculate diff between location and step to get correct direction

**Reply by liodali** [reply]
for polylines you should have offline map  that contain routes (nodes) and you need to implement routing alog like djikestra to find specific route

**Reply by jesussmile** [reply]
The routing website in your plugin gives one-time navigation, I took their coordinates, comparing latitudes I know whether to go right or left, longitude gives me forward or back. I am using my heading to check all these. Then I simply compared my position to those coordinates when I am approaching it. I don't know if it's the right way but it does work for me. I am sure you will do it better. I only used my aviation knowlege.
Here `PointALat` and `PointBLat` are coordinates that I compute to be around 50 meters ahead of the turn. A basic sample
```ruby
Geolocator.getPositionStream().listen((Position position) async {
 if (double.parse((position.latitude).toStringAsFixed(4)) > PointALat&&
          double.parse((position.longitude).toStringAsFixed(4)) < PointBLon) {
        setState(() {
          navigation = "Turn Left";
        });
      }
}
```

**Reply by liodali** [reply]
ok because i'm planning to add those information to drawRoad 
also the problem in turn by turn navigation that i should know if the user still in the same route or not that's what make it hard and also the movement of the user not line that's why we need to use valhalla project to fine tune the user path and detect the new route  if the user change his path like in google maps but for what you did it can be good step to implement this feature

**Reply by liodali** [reply]
i found in orsm that they have match api like in valhalla project to match between user gps and route 
the only problem that i need to make check every minute (for open source project will not be problem)
i will make more research about and i will see how i will start do thing about this feature but thanks for your help

**Reply by jesussmile** [reply]
Sure not a problem.  If I can improve on this I will let you know.

**Comment by felipebueno** [comment]
Hi, @jesussmile!

Can you provide an example app with your implementation of turn-by-turn navigation, please? :slightly_smiling_face:

**Reply by jesussmile** [reply]
> Hi, @jesussmile!
> 
> Can you provide an example app with your implementation of turn-by-turn navigation, please? 🙂


It's not exactly a turn by turn navigation like you see on google map, it's just an example, what I have done is put a point along the path (lat and long), when the position of the driver is greater than that lat-long then it will say turn left.  that's all.\

I am sorry it must not be what you expected but I am working on it, if anything comes thru I will let you know, This can be done by getting the Geo points off the road, So you compare your Geopoint with the next Geo point of the road so you know when to turn left or right or straight, etc.
 right now I am not working on it.

**Comment by KeshavPai** [comment]
> Here's what I have done. Downloaded all the JSON data that contains the API then using some algorithm calculated the latitude and longitude values for all the turns and directions and compared them with my current position, I kept a certain value like 50 or 100 meters to warn the users about changes in direction. I streamed my position to always be in the center of the map with heading and speed. With heading, I can change the orientation of the map to show the road straight ahead. I get a very rough turn-by-turn implementation. Do you think this approach is okay? In the future, I would like to see how you have implemented the polylines and try to compute the directions offline. It's funny how the arrow is dragging my car! 😄
> 
>  My.Video.mp4

Can you share the github repo link for the same if you don't mind. I have to just do a small demo regarding this navigation using OSM. I am totally new to Flutter and OSM as well. I just have to create a small demo of the navigation part and then the rest of the complete implementation will be taken over by the Development Team

**Comment by jesussmile** [comment]
My project is all tied to firebase , it will be difficult for me to create just that and not sure if you will understand.  The basic idea is to get the list of names in the road then compare one point to the another,
The best option is to send the navigation to Google map from your app. For now.

``` ruby
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  double _latitude1 = 37.4219999;
  double _longitude1 = -122.0840575;
  double _latitude2 = 37.7749295;
  double _longitude2 = -122.4194155;

  void _launchMap() async {
    String googleUrl = 'https://www.google.com/maps/dir/?api=1&origin=$_latitude1,$_longitude1&destination=$_latitude2,$_longitude2';
    if (await canLaunch(googleUrl)) {
      await launch(googleUrl);
    } else {
      throw 'Could not open the map.';
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Google Map Navigation'),
      ),
      body: Center(
        child: RaisedButton(
          onPressed: _launchMap,
          child: Text('Launch Map'),
        ),
      ),
    );
  }
}
```
I am sure the author of this plugin will add this feature in future. For now. You can use osm map for everything else theb send navigation to Google maps.

**Comment by liodali** [comment]
I want to share this package : https://github.com/liodali/OSM-Routing-Client-Dart where is for OSRM APIs
and in latest version we added news apis for turn by turn navigation
where you can use `isOnPath` : to check if the user still on the road or not
and `nextInstruction`to get current instruction and next one
we will try to provide example in our plugin here 
and also we will use this package in the future for get road (draw of polyline will be internally )  instead of impl  OSRM for each platform
